src/connector/src/source/base.rs

```diff
@@ -383,16 +384,18 @@ impl ConnectorProperties {
     /// `deny_unknown_fields`: Since `WITH` options are persisted in meta, we do not deny unknown fields when restoring from
     /// existing data to avoid breaking backwards compatibility. We only deny unknown fields when creating new sources.
     pub fn extract(
-        mut with_properties: BTreeMap<String, String>,
+        with_properties: WithOptionsSecResolved,
         deny_unknown_fields: bool,
     ) -> Result<Self> {
-        let connector = with_properties
+        let (options, secret_refs) = with_properties.into_parts();
+        let mut options_with_secret = SECRET_MANAGER.fill_secrets(options, secret_refs)?;
```

Comment:
The logic of calling `fill_secrets` is everywhere - in frontend nodes, compute nodes, connector nodes. How about doing this in Meta node **before** the RPC call? In this way, we don't need to introduce the new field `secret_refs` in these RPC messages, and other nodes will be unaware of secret filling because Meta has done the work for them, making the code less invasive.

Reply:
> How about doing this in Meta node before the RPC call?

The problem is `fill_secrets` is needed even before anything reaching to meta. Examples are we need to fetch iceberg catalog and schema registry in frontend.  

> In this way, we don't need to introduce the new field secret_refs in these RPC messages
 
It's unlikely we can do this. We need to store stream nodes in meta without revealing the real secrets (or it's subject to leak). Thus we have to put in a new filed `secret_refs`

> Meta has done the work for them, making the code less invasive

Meta can build new actors (or stream node) in many places including newly created job, recovery, scaling, altering etc. It's hard to make sure we won't miss any part. Also there are only 10 `fill_secrets` calls, and the use of type `SecretRef` and `WithOptionsSecResolved` make sure we won't miss anything. If we fill the secret somewhere in meta, we can never know if the properties include the secrets or not. (it's map<string, string>)

Also distribute secret is necessary in case of file secret. After all we can't just send path with stream nodes to the cn, we need to distribute the actual secret files

Reply:
Thanks for the explanation. That's convincing.

Reply:
Thoughts: It seems the major cause is because we reused these proto messages for both RPC and persistance. Persistence in meta store requires the secrets must be a reference (`secret_ref`), however, this also cause the RPC to pass a `secret_ref` and the compute nodes therefore have to `fill_secrets` by themselves. This is somehow unexpected but acceptabled to me.